src/Microsoft.TemplateEngine.Cli/New3Command.cs

```diff
@@ -508,6 +522,65 @@ private bool ConfigureLocale()
             return true;
         }
 
+        private bool SyncOptionalWorkloads()
+        {
+            bool isHiveUpdated = false;
+            bool isCustomHive = _commandInput.HasDebuggingFlag("--debug:ephemeral-hive") || _commandInput.HasDebuggingFlag("--debug:custom-hive");
+
+            if (!isCustomHive)
+            {
+                string sdkVersion = EnvironmentSettings.Host.Version;
+
+                try
+                {
+                    List<InstallationRequest> owInstallationRequests = new List<InstallationRequest>();
+                    Dictionary<string, string> owInstalledPkgs = new Dictionary<string, string>();  // packageId -> packageVersion
+                    TemplateLocator optionalWorkloadLocator = new TemplateLocator();
+
+                    IReadOnlyCollection<IOptionalSdkTemplatePackageInfo> owPkgsToSync = optionalWorkloadLocator.GetDotnetSdkTemplatePackages(sdkVersion);
+
+                    foreach (IInstallUnitDescriptor descriptor in _settingsLoader.InstallUnitDescriptorCache.Descriptors.Values)
+                    {
+                        if (descriptor.IsPartOfAnOptionalWorkload)
+                        {
+                            string pkgVersion = "";
+                            if (descriptor.Details.ContainsKey("Version"))
+                            {
+                                pkgVersion = descriptor.Details["Version"];
+                            }
+                            owInstalledPkgs.Add(descriptor.Identifier, pkgVersion);
+                        }
+                    }
+
+                    foreach (IOptionalSdkTemplatePackageInfo packageInfo in owPkgsToSync)
+                    {
+                        if (!owInstalledPkgs.ContainsKey(packageInfo.TemplatePackageId) ||
+                            owInstalledPkgs[packageInfo.TemplatePackageId] != packageInfo.TemplateVersion)
```

Comment:
tiny nit: Similar to above, this can be done with one lookup:

```C#
if (!owInstalledPkgs.TryGetValue(packageInfo.TemplatePackageId, out string version) ||
    version != packageInfo.TemplateVersion) {
```

but if you find the current code more readable I think it's fine to forgo the tiny perf win.